adal/src/main/java/com/microsoft/aad/adal/Oauth2.java

```diff
@@ -474,7 +491,26 @@ private AuthenticationResult postMessage(String requestMessage, Map<String, Stri
                 // Protocol related errors will read the error stream and report
                 // the error and error description
                 Logger.v(TAG, "Token request does not have exception");
-                result = processTokenResponse(response);
+                try {
+                    result = processTokenResponse(response);
+                } catch (final AuthenticationException e) {
+                    if (e.getCode().equals(ADALError.SERVER_NOT_RESPONDING) && retryOnce) {
+                        //retry once if it is a server error
+                        //500, 503 and 504 are the ones we retry
+                        retryOnce = false;
+                        try {
+                            //retry once after half second
+                            final CountDownLatch signal = new CountDownLatch(1);
+                            signal.await(delayTimePeriod, TimeUnit.MILLISECONDS);
+                            Logger.v(TAG, "WebResponse is not a success due to :-" + response.getStatusCode() + "Retrying one more time..");
+                            return postMessage(requestMessage, headers);
+                        } catch (final InterruptedException exception) {
```

Comment:
Wrap the exception specifically around where the exception is thrown. 
Also, I would prefer to just use thread.sleep in this case. No need to use countdownlatch.